File: web/src/services/TimerManager.ts
import {
  TimerState,
  TimerMode,
  PomodoroPhase,
  createInitialTimerState,
  getTotalDuration,
  nextPhase,
  resetTimer,
} from '../models/TimerState';
import { VoiceAlert, shouldTriggerAlert } from '../models/VoiceAlert';
import { AppSettings } from '../models/Settings';
import { ElevenLabsService } from './ElevenLabsService';
import { SettingsManager } from './SettingsManager';

const STATE_KEY = 'pomodoro_voice_timerstate';

export class TimerManager {
  private state: TimerState;
  private timerInterval: number | null = null;
  private triggeredAlerts: Set<string> = new Set();
  private settingsManager: SettingsManager | null = null;
  private elevenLabsService: ElevenLabsService;
  private listeners: Set<() => void> = new Set();

  constructor() {
    this.state = this.loadState();
    this.elevenLabsService = new ElevenLabsService();
    this.requestNotificationPermission();
  }

  setSettingsManager(manager: SettingsManager): void {
    this.settingsManager = manager;
  }

  getState(): TimerState {
    return { ...this.state };
  }

  subscribe(listener: () => void): () => void {
    this.listeners.add(listener);
    return () => {
      this.listeners.delete(listener);
    };
  }

  private notify(): void {
    this.listeners.forEach((listener) => listener());
  }

  private loadState(): TimerState {
    try {
      const stored = localStorage.getItem(STATE_KEY);
      if (stored) {
        return JSON.parse(stored);
      }
    } catch (error) {
      console.error('Error loading timer state:', error);
    }
    return createInitialTimerState();
  }

  private saveState(): void {
    try {
      localStorage.setItem(STATE_KEY, JSON.stringify(this.state));
    } catch (error) {
      console.error('Error saving timer state:', error);
    }
  }

  private async requestNotificationPermission(): Promise<void> {
    if ('Notification' in window && Notification.permission === 'default') {
      await Notification.requestPermission();
    }
  }

  start(): void {
    if (this.state.isRunning) {
      return;
    }

    this.state.isRunning = true;
    this.triggeredAlerts.clear();
    this.saveState();
    this.notify();
    this.startTimer();
  }

  pause(): void {
    this.state.isRunning = false;
    if (this.timerInterval !== null) {
      clearInterval(this.timerInterval);
      this.timerInterval = null;
    }
    this.saveState();
    this.notify();
  }

  reset(): void {
    this.pause();
    this.state = resetTimer(this.state);
    this.triggeredAlerts.clear();
    this.saveState();
    this.notify();
  }

  skip(): void {
    this.pause();
    this.state = nextPhase(this.state);
    this.triggeredAlerts.clear();
    this.saveState();
    this.checkAndTriggerAlerts(true);
    this.notify();
    if (this.state.isRunning) {
      this.startTimer();
    }
  }

  setCustomDuration(minutes: number): void {
    this.state.customDuration = minutes * 60;
    if (this.state.mode === TimerMode.Custom) {
      this.state.remainingSeconds = this.state.customDuration;
    }
    this.saveState();
    this.notify();
  }

  setMode(mode: TimerMode): void {
    this.pause();
    this.state.mode = mode;
    this.triggeredAlerts.clear();
    if (mode === TimerMode.Custom) {
      this.state.remainingSeconds = this.state.customDuration;
    } else {
      this.state = resetTimer(this.state);
    }
    this.saveState();
    this.notify();
  }

  private startTimer(): void {
    if (this.timerInterval !== null) {
      clearInterval(this.timerInterval);
    }

    this.timerInterval = window.setInterval(() => {
      this.tick();
    }, 1000);
  }

  private tick(): void {
    if (!this.state.isRunning) {
      return;
    }

    this.checkAndTriggerAlerts(false);

    if (this.state.remainingSeconds > 0) {
      this.state.remainingSeconds -= 1;
      this.saveState();
      this.notify();
    } else {
      this.timerCompleted();
    }
  }

  private checkAndTriggerAlerts(isTransitioning: boolean): void {
    if (!this.settingsManager) {
      return;
    }

    const settings = this.settingsManager.getSettings();
    const alerts: VoiceAlert[] =
      this.state.phase === PomodoroPhase.Focus
        ? settings.voiceAlerts.focusAlerts
        : settings.voiceAlerts.breakAlerts;

    for (const alert of alerts) {
      if (this.triggeredAlerts.has(alert.id)) {
        continue;
      }

      if (
        shouldTriggerAlert(
          alert,
          this.state.remainingSeconds,
          getTotalDuration(this.state),
          this.state.phase,
          isTransitioning
        )
      ) {
        this.triggeredAlerts.add(alert.id);
        this.triggerAlert(alert);
      }
    }
  }

  private async triggerAlert(alert: VoiceAlert): Promise<void> {
    if (!this.settingsManager) {
      return;
    }

    const settings = this.settingsManager.getSettings();
    if (
      !settings.elevenLabsAPIKey ||
      !settings.voiceID ||
      !alert.message
    ) {
      return;
    }

    try {
      const data = await this.elevenLabsService.synthesizeSpeech(
        alert.message,
        settings.voiceID,
        settings.elevenLabsAPIKey,
        settings.volume
      );
      await this.elevenLabsService.playAudio(data, settings.volume);
    } catch (error) {
      console.error('Voice alert error:', error);
    }
  }

  private timerCompleted(): void {
    this.pause();

    // Check "end of session" alerts before transitioning
    this.checkAndTriggerAlerts(false);

    // Show notification
    if ('Notification' in window && Notification.permission === 'granted') {
      const title =
        this.state.phase === PomodoroPhase.Focus
          ? 'Focus Session Complete'
          : this.state.phase === PomodoroPhase.ShortBreak
          ? 'Short Break Complete'
          : 'Long Break Complete';

      new Notification(title, {
        body: 'Timer completed!',
        icon: '/vite.svg',
      });
    }

    // Auto-advance to next phase
    if (this.state.mode === TimerMode.Pomodoro) {
      this.state = nextPhase(this.state);
      this.triggeredAlerts.clear();
      // Check "start of break" alerts after transitioning
      this.checkAndTriggerAlerts(true);
    } else {
      this.state.remainingSeconds = getTotalDuration(this.state);
      this.triggeredAlerts.clear();
    }

    this.saveState();
    this.notify();
  }
}

File: Services/TimerManager.swift
//
//  TimerManager.swift
//  PomodoroVoice
//

import Foundation
import UserNotifications
import Combine

class TimerManager: ObservableObject {
    @Published var state: TimerState {
        didSet {
            saveState()
        }
    }
    
    @Published var isTransitioning: Bool = false
    
    private var timer: Timer?
    private let stateKey = "com.pomodorovoice.timerstate"
    private let elevenLabsService = ElevenLabsService()
    private var settingsManager: SettingsManager?
    private var triggeredAlerts: Set<UUID> = [] // Track which alerts have been triggered in current session
    
    init() {
        if let data = UserDefaults.standard.data(forKey: stateKey),
           let decoded = try? JSONDecoder().decode(TimerState.self, from: data) {
            self.state = decoded
        } else {
            self.state = TimerState()
        }
        
        requestNotificationPermission()
    }
    
    func setSettingsManager(_ manager: SettingsManager) {
        self.settingsManager = manager
    }
    
    private func saveState() {
        if let encoded = try? JSONEncoder().encode(state) {
            UserDefaults.standard.set(encoded, forKey: stateKey)
        }
    }
    
    private func requestNotificationPermission() {
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound]) { granted, error in
            if let error = error {
                print("Notification permission error: \(error)")
            }
        }
    }
    
    func start() {
        guard !state.isRunning else { return }
        state.isRunning = true
        triggeredAlerts.removeAll() // Reset triggered alerts when starting
        startTimer()
    }
    
    func pause() {
        state.isRunning = false
        timer?.invalidate()
        timer = nil
    }
    
    func reset() {
        pause()
        state.reset()
        triggeredAlerts.removeAll()
    }
    
    func skip() {
        pause()
        state.nextPhase()
        triggeredAlerts.removeAll() // Reset for new phase
        isTransitioning = true
        checkAndTriggerAlerts(isTransitioning: true)
        isTransitioning = false
        if state.isRunning {
            startTimer()
        }
    }
    
    func setCustomDuration(minutes: Int) {
        state.customDuration = minutes * 60
        if state.mode == .custom {
            state.remainingSeconds = state.customDuration
        }
    }
    
    func setMode(_ mode: TimerMode) {
        pause()
        state.mode = mode
        triggeredAlerts.removeAll()
        if mode == .custom {
            state.remainingSeconds = state.customDuration
        } else {
            state.reset()
        }
    }
    
    private func startTimer() {
        timer?.invalidate()
        timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            self?.tick()
        }
        RunLoop.current.add(timer!, forMode: .common)
    }
    
    private func tick() {
        guard state.isRunning else { return }
        
        checkAndTriggerAlerts(isTransitioning: false)
        
        if state.remainingSeconds > 0 {
            state.remainingSeconds -= 1
        } else {
            timerCompleted()
        }
    }
    
    private func checkAndTriggerAlerts(isTransitioning: Bool) {
        guard let settings = settingsManager?.settings else { return }
        
        let alerts: [VoiceAlert]
        switch state.phase {
        case .focus:
            alerts = settings.voiceAlerts.focusAlerts
        case .shortBreak, .longBreak:
            alerts = settings.voiceAlerts.breakAlerts
        }
        
        for alert in alerts {
            // Skip if already triggered in this session
            guard !triggeredAlerts.contains(alert.id) else { continue }
            
            if alert.shouldTrigger(
                currentSeconds: state.remainingSeconds,
                totalSeconds: state.totalDuration,
                phase: state.phase,
                isTransitioning: isTransitioning
            ) {
                triggeredAlerts.insert(alert.id)
                triggerAlert(alert)
            }
        }
    }
    
    private func triggerAlert(_ alert: VoiceAlert) {
        guard let settings = settingsManager?.settings,
              !settings.elevenLabsAPIKey.isEmpty,
              !settings.voiceID.isEmpty,
              !alert.message.isEmpty else { return }
        
        elevenLabsService.stopAudio() // Stop any currently playing audio
        elevenLabsService.synthesizeSpeech(
            text: alert.message,
            voiceID: settings.voiceID,
            apiKey: settings.elevenLabsAPIKey,
            volume: settings.volume
        ) { [weak self] result in
            DispatchQueue.main.async {
                switch result {
                case .success(let data):
                    self?.elevenLabsService.playAudio(data: data, volume: settings.volume)
                case .failure(let error):
                    print("Voice alert error: \(error)")
                }
            }
        }
    }
    
    private func timerCompleted() {
        pause()
        
        // Check "end of session" alerts before transitioning
        isTransitioning = false
        checkAndTriggerAlerts(isTransitioning: false)
        
        // Show notification
        let content = UNMutableNotificationContent()
        content.title = phaseTitle
        content.body = "Timer completed!"
        content.sound = .default
        
        let request = UNNotificationRequest(
            identifier: UUID().uuidString,
            content: content,
            trigger: nil
        )
        
        UNUserNotificationCenter.current().add(request)
        
        // Auto-advance to next phase
        if state.mode == .pomodoro {
            state.nextPhase()
            triggeredAlerts.removeAll() // Reset for new phase
            // Check "start of break" alerts after transitioning
            isTransitioning = true
            checkAndTriggerAlerts(isTransitioning: true)
        } else {
            state.remainingSeconds = state.totalDuration
            triggeredAlerts.removeAll() // Reset for new cycle
        }
        
        isTransitioning = false
    }
    
    private var phaseTitle: String {
        switch state.phase {
        case .focus:
            return "Focus Session Complete"
        case .shortBreak:
            return "Short Break Complete"
        case .longBreak:
            return "Long Break Complete"
        }
    }
    
    deinit {
        timer?.invalidate()
    }
}

